lib/room.cpp

```diff
@@ -1430,8 +1494,17 @@ void Room::handleRoomKeyEvent(const RoomKeyEvent& roomKeyEvent,
     }
     if (d->addInboundGroupSession(senderKey, roomKeyEvent.sessionId(),
                                   roomKeyEvent.sessionKey())) {
-        qCDebug(E2EE) << "added new inboundGroupSession:"
-                      << d->groupSessions.count();
+        qCWarning(E2EE) << "added new inboundGroupSession:"
+                      << d->groupSessions.size();
+        for (unsigned long int i = 0; i < d->timeline.size(); i++) {
+            if (auto encryptedEvent = d->timeline[i].viewAs<EncryptedEvent>()) {
+                auto decrypted = decryptMessage(*encryptedEvent);
+                if(decrypted) {
+                    auto oldEvent = d->timeline[i].replaceEvent(std::move(decrypted));
+                    emit replacedEvent(d->timeline[i].event(), rawPtr(oldEvent));
+                }
+            }
+        }
```

Comment:
I have a few problems with this code. First, AIUI, it rolls over potentially thousands of events every time a new session is around. Second, even if it's preferred to lazier decryption whenever an event needs to be shown, the code uses suboptimal index access although timeline items don't move when events are replaced so you can use normal range iteration. Finally, maybe replacing the event is not the best way overall since you lose the original encrypted event and can no more, e.g. verify the sender. Correct me if I'm wrong.

Reply:
I'm not happy with it either for mostly the same reasons.

- We can make it a bit faster by checking the sessionId before trying to decrypt, but it is indeed a bit ugly
- I don't quite understand your second point
- Replacing the event completely is quite nice since it works completely without the client needing to have any special handling for encrypted events. The point about keeping around the encrypted event is indeed important, maybe we can keep this around in an additional field in timelineitem or event?